File: max_dip_count.rb
# frozen_string_literal: true

# Function to calculate the total number of times the stamp was dipped in the ink
def calculate_dip_count(stamp_times, ink_duration)
  dip_count = 0
  current_dip_time = 0

  stamp_times.each do |stamp_time|
    if stamp_time.zero? || stamp_time > current_dip_time
      dip_count += 1
      current_dip_time = stamp_time + ink_duration
    end
  end

  dip_count
end

# Input: Number of stamps, Stamp times, and Ink duration
puts 'Enter the number of stamps:'
n = gets.to_i

if n.zero?
  puts 0
  exit
end

puts 'Enter the stamp times (sorted, space-separated):'
stamp_times = gets.split.map(&:to_i)

puts 'Enter the ink duration:'
ink_duration = gets.to_i

if ink_duration.zero?
  puts n
  exit
end

# Calculate and output the dip count
result = calculate_dip_count(stamp_times, ink_duration)
puts "The total number of times the stamp was dipped in the ink: #{result}"

File: max_dip_count/solutions/max_dip_count.c
#include <stdio.h>

// function to calculate the total number of times the stamp was dipped in the ink pad
int calculate_dip_count(int stamp_times[], int n, int ink_duration)
{
    int dip_count = 0;
    int current_dip_time = 0;
    for (int i = 0; i < n; i++)
    {
        if (stamp_times[i] == 0 || stamp_times[i] > current_dip_time)
        {
            dip_count++;
            current_dip_time = stamp_times[i] + ink_duration;
        }
    }
    return dip_count;
}

int main()
{

    int n;
    scanf("%d", &n);

    if (n == 0)
    {
        printf("%d", n);
        return 0;
    }

    int stamp_times[n];
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &stamp_times[i]);
    }

    int ink_duration;
    scanf("%d", &ink_duration);

    if (ink_duration == 0)
    {
        printf("%d", n);
        return 0;
    }

    printf("%d", calculate_dip_count(stamp_times, n, ink_duration));
}